Added a custom JVP rule for jax.nn.logsumexp

superbobry opened the pull request:
Fixes #22398 where the Jacobian of jax.nn.logsumexp was wrong if b= contained exact zeros.

superbobry committed:
Added a custom JVP rule for jax.nn.logsumexp

Fixes #22398 where the Jacobian of jax.nn.logsumexp was wrong if b= contained
exact zeros.

superbobry commented:
@jakevdp does this look like the fix you had in mind?

dfm commented on jax/_src/ops/special.py:
```diff
@@ -95,3 +97,22 @@ def logsumexp(a: ArrayLike, axis: Axis = None, b: ArrayLike | None = None,
+      logsumexp(a, axis, b * a_dot + b_dot, keepdims, where) - out)
+
+  if return_sign:
+    return (out, sign), out_dot
```
I think this probably needs a `sign_dot` here too. I think the tangents always need to have the same pytree structure as the primals.

superbobry replied:
I could be wrong, but I think `sign_dot` is included in `out_dot` if `return_sign` is True.

superbobry replied:
Added a test with `return_sign=True`.

dfm commented on jax/_src/ops/special.py:
```diff
@@ -95,3 +97,22 @@ def logsumexp(a: ArrayLike, axis: Axis = None, b: ArrayLike | None = None,
+  if return_sign:
+    out, sign = out
+
+  if b is None:
+    b = jnp.ones_like(a)
```
We probably need to handle the case where `b_dot` is `None` as well?

superbobry replied:
Good point. Am I right that `b_dot` is only None if `b` is None?

superbobry replied:
Done.

jakevdp commented on jax/_src/ops/special.py:
```diff
@@ -95,3 +97,22 @@ def logsumexp(a: ArrayLike, axis: Axis = None, b: ArrayLike | None = None,
+  if b is None:
+    b = jnp.ones_like(a)
+  out_dot = jnp.exp(
+      logsumexp(a, axis, b * a_dot + b_dot, keepdims, where) - out)
```
This expression looks a bit strange to me – each element of `out_dot` should reflect the gradient of the corresponding elements of `a_dot` and `b_dot`, so summing over them doesn't look right to me.

superbobry replied:
I think we do need to sum along `axis`, otherwise the shapes don't add up: `b * a_dot + b_dot` has the same shape as `a`, but the expected `out_dot` shape should be the same as `out`, i.e. lower rank than `a`.

jakevdp commented:
Let's double-check the formula here. I get this:

$$f(a, b) = \log\sum_i b_i \exp(a_i)$$

$$df_j = da_j \ \partial f/\partial a_j + db_j \ \partial f/\partial b_j$$

$$df_j = da_j b_j\exp(a_j)/\sum_i b_i \exp(a_i) + db_j \exp(a_j)/\sum_i b_i\exp(a_i)$$

$$df_j = (da_j b_j + db_j)\exp(a_j)/\exp(\log\sum_i b_i\exp(a_i))$$

$$df_j = (da_j b_j + db_j)\exp[a_j - f(a, b)]$$

So the out tangent in the general case should be something like
```python
(a_dot * b + b_dot) * jnp.exp(a - out)
```
Does that sound right?

superbobry commented on jax/_src/ops/special.py:
```diff
@@ -95,3 +97,22 @@ def logsumexp(a: ArrayLike, axis: Axis = None, b: ArrayLike | None = None,
+def _logsumexp_jvp(axis, keepdims, return_sign, where, primals, tangents):
+  a, b = primals
+  a_dot, b_dot = tangents
+  out = logsumexp(a, axis, b, keepdims, return_sign, where)
```
Done.

superbobry commented:
Thanks, Jake! I think this looks right, but it also seems that we need to sum along `axis` to match the `out` shape, which is why I had a `logsumexp`. Wdyt?

dfm reviewed (approved):
Looks good!

jakevdp commented:
> Thanks, Jake! I think this looks right, but it also seems that we need to sum along `axis` to match the `out` shape, which is why I had a `logsumexp`. Wdyt?

Ah, right. I think `sum` is what we want here. The output tangent should be linear in the input tangents, so taking the exponent and log of the input tangents doesn't sound right.

superbobry commented on tests/lax_scipy_test.py:
```diff
@@ -202,6 +203,34 @@ def testLogSumExpWhere(self, shape, dtype):
+      axis=[None, 0], with_b=[False, True], return_sign=[False, True]
+  )
+  def testLogSumExpJac(self, axis, with_b, return_sign):
+    # TODO(slebedev): Test return_sign=True.
```
Done.

superbobry commented:
Discussed with Jake offline:

* We can use `exp(logsumexp(...))`, since exp/log cancel out and the `out_dot` is still linear in the inputs, but
* it is somewhat of a code smell to use non-linear operations on tangents;
* we will get nans if tangents are negative with `logsumexp`.

So, I switched to `sum` and also removed unnecessary allocations for the case of `b is None`.

jakevdp reviewed (approved):
Looks good!
